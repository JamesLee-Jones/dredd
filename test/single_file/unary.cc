int main() {
  int x = 3;
  x++;
  ++x;
  --x;
  x--;
  return -x;
}
