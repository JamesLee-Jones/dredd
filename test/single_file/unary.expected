#include <cinttypes>
#include <cstddef>
#include <functional>
#include <string>

static bool __dredd_enabled_mutation(int local_mutation_id) {
  static bool initialized = false;
  static uint64_t enabled_bitset[2];
  if (!initialized) {
    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable != nullptr) {
      std::string contents(dredd_environment_variable);
      while (true) {
        size_t pos = contents.find(",");
        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
        if (!token.empty()) {
          int value = std::stoi(token);
          int local_value = value - 0;
          if (local_value >= 0 && local_value < 68) {
            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
          }
        }
        if (pos == std::string::npos) {
          break;
        }
        contents.erase(0, pos + 1);
      }
    }
    initialized = true;
  }
  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
}

static int __dredd_replace_unary_operator_PreInc_int(std::function<int&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg();
  return ++arg();
}

static int __dredd_replace_unary_operator_PreDec_int(std::function<int&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg();
  return --arg();
}

static int __dredd_replace_unary_operator_PostInc_int(std::function<int&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg();
  return arg()++;
}

static int __dredd_replace_unary_operator_PostDec_int(std::function<int&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg();
  return arg()--;
}

static int __dredd_replace_unary_operator_Minus_int(std::function<int&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg();
  return -arg();
}

static float __dredd_replace_unary_operator_PreInc_float(std::function<float&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return arg();
  return ++arg();
}

static float __dredd_replace_unary_operator_PreDec_float(std::function<float&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return arg();
  return --arg();
}

static float __dredd_replace_unary_operator_PostInc_float(std::function<float&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return arg();
  return arg()++;
}

static float __dredd_replace_unary_operator_PostDec_float(std::function<float&()> arg, int local_mutation_id) {
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++arg();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return --arg();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return +arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return arg();
  return arg()--;
}

int main() {
  int x = 3;
  float y = 3.532;
  if (!__dredd_enabled_mutation(59)) { __dredd_replace_unary_operator_PostInc_int([&]() -> int& { return static_cast<int&>(x); }, 0); }
  if (!__dredd_enabled_mutation(60)) { __dredd_replace_unary_operator_PostInc_float([&]() -> float& { return static_cast<float&>(y); }, 7); }
  if (!__dredd_enabled_mutation(61)) { __dredd_replace_unary_operator_PreInc_int([&]() -> int& { return static_cast<int&>(x); }, 13); }
  if (!__dredd_enabled_mutation(62)) { __dredd_replace_unary_operator_PreInc_float([&]() -> float& { return static_cast<float&>(y); }, 20); }
  if (!__dredd_enabled_mutation(63)) { __dredd_replace_unary_operator_PreDec_int([&]() -> int& { return static_cast<int&>(x); }, 26); }
  if (!__dredd_enabled_mutation(64)) { __dredd_replace_unary_operator_PreDec_float([&]() -> float& { return static_cast<float&>(y); }, 33); }
  if (!__dredd_enabled_mutation(65)) { __dredd_replace_unary_operator_PostDec_int([&]() -> int& { return static_cast<int&>(x); }, 39); }
  if (!__dredd_enabled_mutation(66)) { __dredd_replace_unary_operator_PostDec_float([&]() -> float& { return static_cast<float&>(y); }, 46); }
  if (!__dredd_enabled_mutation(67)) { return __dredd_replace_unary_operator_Minus_int([&]() -> int& { return static_cast<int&>(x); }, 52); }
}
